src/activatable/activatable.ts

```diff
@@ -0,0 +1,130 @@
+  [LifeCycleState.Deactivating]: 'is deactivating',
+  [LifeCycleState.Deactivated]: 'is deactivated',
+  [LifeCycleState.Disposed]: 'is disposed',
+  [LifeCycleState.FailedToActivate]: 'errored while activating',
+  [LifeCycleState.FailedToDeactivate]: 'errored while deactivating',
+};
+
+export interface IActivatable {
+  isActivating: boolean;
+  isActivated: boolean;
+  isDeactivating: boolean;
+  isDeactivated: boolean;
+  isErrored: boolean;
+  activate: () => Promise<void>;
+  deactivate: () => Promise<void>;
+  disposeActivateable: () => void;
+}
+
+export abstract class Activatable implements IActivatable {
+  private _state: LifeCycleState = LifeCycleState.Created;
+
+  public get isActivating() {
+    return this._state === LifeCycleState.Activating;
+  }
+
+  public get isActivated() {
+    return this._state === LifeCycleState.Activated;
+  }
+
+  public get isDeactivating() {
+    return this._state === LifeCycleState.Deactivating;
+  }
+
+  public get isDeactivated() {
+    return (
+      this._state === LifeCycleState.Created ||
+      this._state === LifeCycleState.Deactivated ||
+      this._state === LifeCycleState.Disposed
+    );
+  }
+
+  public get isErrored() {
+    return this._state === LifeCycleState.FailedToActivate || this._state === LifeCycleState.FailedToDeactivate;
+  }
+
+  public async activate() {
+    switch (this._state) {
+      case LifeCycleState.Activated:
+        return;
+      case LifeCycleState.Created:
+      case LifeCycleState.Deactivated:
+        try {
+          this._state = LifeCycleState.Activating;
+          await this._activate();
+          this._state = LifeCycleState.Activated;
+          return;
+        } catch (e) {
+          this._state = LifeCycleState.FailedToActivate;
+          throw e;
+        }
+      default:
+        throw new Error(`Cannot activate an Activatable that ${LifeCycleDescription[this._state]}`);
+    }
+  }
+
+  public async deactivate() {
+    switch (this._state) {
+      case LifeCycleState.Deactivated:
+        return;
+      case LifeCycleState.Activated:
+        try {
+          this._state = LifeCycleState.Deactivating;
+          await this._deactivate();
+          this._state = LifeCycleState.Deactivated;
+          return;
+        } catch (e) {
+          this._state = LifeCycleState.FailedToDeactivate;
+          throw e;
+        }
+      default:
+        throw new Error(`Cannot deactivate an Activatable that ${LifeCycleDescription[this._state]}`);
+    }
+  }
+
+  public disposeActivateable() {
+    this._state = LifeCycleState.Disposed;
+  }
+
+  protected abstract _activate(): void | Promise<void>;
+
+  protected abstract _deactivate(): void | Promise<void>;
+}
+
+export class ActivatableCollection extends Activatable implements IActivatable {
+  protected readonly activatables: IActivatable[] = [];
+
+  protected async _activate() {
+    await Promise.all(this.activatables.map(activatable => activatable.activate()));
```

Comment:
This is tricky. `Promise.all` is very likely to be evil here in case a single `activatable` could not be activated (rejected promise).

I'd suggest placing a try catch there or something.
Alternatively, we could use `Promise.allSettled` (will need polyfilling).

Reply:
I'm pretty sure it's the correct semantics. If not all of the children are activated, then the parent is not activated.

Reply:
> If not all of the children are activated, then the parent is not activated.

This is absolutely fine.

The problem is that we should always until the activation process of each node is fully finished.
To illustrate:
```ts
try {
await Promise.all([
  new Promise((resolve, reject) => {
    reject()
  }),
  new Promise((resolve, reject) => {
    setTimeout(() => {
      // I am stll activating!
      console.log('bam');
      resolve();
   }, 1000)
  });
])
catch {
console.log('woop')
}
```

Reply:
That... poses an interesting problem. And suggests I need to actually label all the edges in my state diagram with which methods correspond to which transition.

Do we want to add `abortActivating` and `abortDeactivating` methods? That makes the logic more complex, but would allow us to cancel activation.

Truth be told I haven't thought through the state diagram for the ActivatableCollection very well. If some of the children fail to activate but some succeed, should we then automatically deactivate the activated ones so the collection as a whole is in a `deactivated/FailedToActivate` state? Ugh. what if deactivating some then fail, then we have a mixed `FailedToDeactivate/FailedToActivate/deactivated` state.

Reply:
> Do we want to add abortActivating and abortDeactivating methods? That makes the logic more complex, but would allow us to cancel activation.

I think we should be okay with not having them. Can't seem to find any use cases for now.



> f some of the children fail to activate but some succeed, should we then automatically deactivate the activated ones so the collection as a whole is in a deactivated/FailedToActivate state? Ugh. what if deactivating some then fail, then we have a mixed FailedToDeactivate/FailedToActivate/deactivated state.

This is exactly what I was trying to say.
I *think* we should `deactivate` all other children if one of the child refused to activate, **but** this could lead to a scenario where one child permanently blocks the parent if it fails to activate all the time.

Reply:
The other possibility... and as sub-optimal as it is, it might be the right one (for reasons I'll list below) is to activate them in series.

```
for (let child of this._activatables) {
  await child.activate()
}
```

pros:
1. we do this already anyway:

https://github.com/stoplightio/studio-internal/blob/0692ccf4cf4049402db52e2eddf285eda4975a53/src/stores/project/store.ts#L164-L168

2. It might spare us from stupid race conditions

3. It leaves us in a more predictable state. If `activate` fails we know that the state of the children will exactly be `[Activated*, FailedToActivate, Deactivated*]` instead of a weird mix of `Activated`, `Activating`, `FailedToActivate`, and `Deactivated`.

4. It makes performance profiling simpler. We'll know which `activate` takes the longest.

cons:
1. it might be perceptibly slower. Not slower than it _currently_ is (since we already activate things in serial) but slower than it _could_ be.

Reply:
> It might spare us from stupid race conditions

:+1:, but in general, so does `Promise.allSettled`. `activate` will always be called in the same order, which is our only concern.

> It leaves us in a more predictable state. If activate fails we know that the state of the children will exactly be [Activated*, FailedToActivate, Deactivated*] instead of a weird mix of Activated, Activating, FailedToActivate, and Deactivated.


That's right, but yet again, `Promise.allSettled` has us covered.


How do you feel about trying out my initial suggestion?
> I'd suggest placing a try catch there or something.

```ts
await Promise.all(this.activatables.map(async activatable => {
  try {
     await activatable.activate();
     return LifeCycleState.Activated;
  } catch {
     return LifeCycleState.FailedToActivate;
  }
}));
```

Reply:
😕  `Promise.allSettled` doesn't evaluate them serially. So I think you're not thinking of the same stupid race conditions. I'm thinking of far stupider ones (aka bugs in our store hierarchy with implicit dependencies that really should be fixed architecturally but don't because they only show up for a small number of customers a small percent of the time).

>>I'd suggest placing a try catch there or something.

Oh like THAT, on the inside, I thought you meant on the outside. That looks very nice. Yeah.

Reply:
I think I'm gonna go for the serial for loop though. It's just much simpler.